### PR TITLE
Add Initial support for Analyzing Indirect Control Flow on Amdgpu Vega

### PR DESCRIPTION
**Pull request by bbiiggppiigg**
This series of commits added the some bare minimum functionalities that are required to resolve one type of indirect control flow that can be observed in kernels of amdgpu, namely functional call performed by the following instruction sequence

getpc
add
addc
swappc

### CODE REVIEW

**Comment by hainest** [comment]
@bbiiggppiigg Could you rebase this onto the head of master?

File: dataflowAPI/src/RoseInsnFactory.C [hunk 1]
```diff
@@ -402,8 +406,191 @@ bool RoseInsnArmv8Factory::handleSpecialCases(entryID, SgAsmInstruction *, SgAsm
...
+                auto [oper0,oper1] = expandSgprPair(operands[0]);
+                operands.resize(3);
+                operands[0] = oper0;
+                operands[1] = oper1;
+                operands[2] = Operand(InstructionAPI::Immediate::makeImmediate(Result(u64,_addr+4)),false,false); 
+
+                break;
+            }
+ 
+        case e_lea: {
```
Comment (mxz297) [inline dataflowAPI/src/RoseInsnFactory.C:493]: @bbiiggppiigg Why the AMD GPU architecture code needs to deal with x86 CPU opcodes?

**Comment by mxz297** [comment]
@bbiiggppiigg Did you ran this PR against the Dyninst test suite on any platform?

**Comment by bbiiggppiigg** [comment]
@mxz297 I wasn't able to get the test suite working.  I'll try to do it.

**Comment by hainest** [comment]
I ran this through my standard testing battery last night. There are several regressions on x86:

https://bottle.cs.wisc.edu/search?dyninst_branch=PR979

I'll re-run the updates you just posted.

**Review by bbiiggppiigg — COMMENTED** [review]
This should be fixed by commit 8c93e13

**Comment by bbiiggppiigg** [comment]
@hainest I've reverted the outputting of those messages back to use parsing_printf. so those messages should no longer appear.

**Comment by hainest** [comment]
Great, thanks! Let me know when you're ready, and I'll re-run the test suite battery.

**Comment by mxz297** [comment]
@hainest Thanks for running the test suite.

@bbiiggppiigg This PR will require more testing. Getting no regression is the first step as your change may have side affects on jump table analysis on CPU. After you get the test regression fixed, we will need to setup some way to compare ParseAPI output on more binaries before this PR can be merged.

**Comment by bbiiggppiigg** [comment]
@mxz297 What you've said totally make sense. 
I'll inspect my commits and see if I can find the problem.

@hainest I think I've removed most debugging messages so they shouldn't break your log parser, but since I haven't find out the cause of regression, I expect the regression to persist after your re-run

**Comment by hainest** [comment]
https://bottle.cs.wisc.edu/search?dyninst_branch=PR979

The logging output has been fixed, so that my build script's parser is working again. ARM sees no issues, but x86 and Power are having regressions.

**Comment by bbiiggppiigg** [comment]
I've tracked down the commit that causes the regression on x86, and but I'm still analyzing the cause

**Comment by hainest** [comment]
https://bottle.cs.wisc.edu/search?dyninst_branch=PR979&since_date=2021-02-07

There are still regressions in test1_33 on Power. The failure on ARM is unrelated to this change.

**Comment by hainest** [comment]
https://bottle.cs.wisc.edu/search?dyninst_branch=PR979&since_date=2021-02-23

No new regressions now.

**Comment by hainest** [comment]
@bbiiggppiigg What's the status of this PR? Do you need to do more testing before merging?

**Comment by bbiiggppiigg** [comment]
The intended functionality of this PR should be working correctly if there isn't any new regression, and for whatever that was working before on amdgpu, this PR shouldn't introduce new bug.  
I'm aware of some potential buffer overflows that needs to be fixed, but that should probably be another PR of its own.

**Comment by hainest** [comment]
@bbiiggppiigg Are you comfortable adding this to the 11.0 release, or would you rather wait until the buffer overflows are fixed?

**Comment by bbiiggppiigg** [comment]
@hainest I think it would be better if these buffer overflows are fixed before they become a part of the stable release. Saying that, the buffer overflows should have already existed before this PR. 
So maybe we can merge this PR, and I'll try to submit a separate PR for fixing buffer overflows by end of this week?

**Comment by mxz297** [comment]
@bbiiggppiigg We need to test this PR a little bit more before merging it.

Here is a test program that uses Dyninst to print CFG: https://github.com/HPCToolkit/hpctoolkit-tests/blob/master/dyninst/openmp-loops.cpp

I would like you to use this program to compare the output of current master and current master with your PR. Ideally, the output should be the same. 

You should be able to run this comparison across all system libraries. Let me know if you need more helps to set up the test.

**Comment by bbiiggppiigg** [comment]
@mxz297 
I've tested on amd2 and there is not difference in output for all binaries under /bin/*
Are there other binaries that I should test for ?

**Comment by mxz297** [comment]
@bbiiggppiigg please also test with binaries under /lib or /lib64. So we need to test both executables and libraries.

In addition to amd2, can you test on at least zeroah (ARM) and a power machine? Do you have access to a Power machine? 

AMD2 has ubuntu 18.04. @hainest Do you other x86-64 candidates? I think we should at least try a redhat system.

**Comment by bbiiggppiigg** [comment]
@mxz297 
On amd2 the output is the same for /lib/x86_64-linux-gnu/* /bin/* /lib32/* and /libx32/*
For power machine I'm planning to test on llnl@rice and I'm setting it up now, but I don't think I have any access to an arm machine, I"ll ask Tim for access to it.

**Comment by bbiiggppiigg** [comment]
@mxz297 @hainest 
Tested on zeroah, llnl and amd2 and Rebased on master

**Comment by mxz297** [comment]
@bbiiggppiigg Thanks!

@hainest You can merge this PR after a full test suite run.

**Comment by hainest** [comment]
https://bottle.cs.wisc.edu/search?dyninst_branch=PR979&since_date=2021-04-04

LGTM.

**Pull request merged**